crates/pixi_build_frontend/src/tool/cache.rs

```diff
@@ -1,38 +1,149 @@
+    }
+
+    pub fn with_cache_dir(mut self, cache_dir: PathBuf) -> Self {
+        self.cache_dir = cache_dir;
+        self
+    }
+
+    pub fn build(self) -> ToolContext {
+        let gateway = Gateway::builder()
+            .with_client(self.client.clone())
+            .with_cache_dir(self.cache_dir.join(CONDA_REPODATA_CACHE_DIR))
+            .with_channel_config(self.gateway_config)
+            .finish();
+
+        ToolContext {
+            channels: self.channels,
+            cache_dir: self.cache_dir,
+            client: self.client,
+            gateway,
+        }
+    }
+}
+
+/// The tool context,
+/// containing client, channels and gateway configuration
+/// that will be used to resolve and install tools.
+#[derive(Default, Clone)]
+pub struct ToolContext {
+    // Authentication client to use for fetching repodata.
+    pub client: ClientWithMiddleware,
+    /// The channels to use for resolving tools.
+    pub channels: Vec<Channel>,
+    // The cache directory to use for the tools.
+    pub cache_dir: PathBuf,
+    // The gateway to use for fetching repodata.
+    pub gateway: Gateway,
+}
+
+impl Debug for ToolContext {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ToolContext")
+            .field("client", &self.client)
+            .field("channels", &self.channels)
+            .field("cache_dir", &self.cache_dir)
+            .finish()
+    }
+}
+
+impl ToolContext {
+    /// Create a new tool context.
+    pub fn new(
+        client: ClientWithMiddleware,
+        gateway: Gateway,
+        cache_dir: PathBuf,
+        channels: Vec<Channel>,
+    ) -> Self {
+        Self {
+            client,
+            channels,
+            cache_dir,
+            gateway,
+        }
+    }
+
+    /// Create a new tool context builder with the given channels.
+    pub fn builder(channels: Vec<Channel>) -> ToolContextBuilder {
+        ToolContextBuilder::new(channels)
+    }
+}
+
 /// A [`ToolCache`] maintains a cache of environments for build tools.
 ///
 /// This is useful to ensure that if we need to build multiple packages that use
 /// the same tool, we can reuse their environments.
+/// (nichita): it can also be seen as a way to create tools itself
+#[derive(Default, Debug)]
 pub struct ToolCache {
-    cache: DashMap<CacheableToolSpec, CachedTool>,
+    /// The cache of tools.
+    pub cache: DashMap<CacheableToolSpec, CachedTool>,
+    /// The context for the tools.
+    /// It contains necessary details
+    /// for the tools to be resolved and installed
+    pub context: ToolContext,
 }
 
 #[derive(thiserror::Error, Debug)]
 pub enum ToolCacheError {
     #[error("could not resolve '{}', {1}", .0.display())]
     Instantiate(PathBuf, which::Error),
+    #[error("could not install isolated tool '{}'", .0.as_display())]
+    Install(miette::Report),
+    #[error("could not determine default cache dir '{}'", .0.as_display())]
+    CacheDir(miette::Report),
```

Comment:
```suggestion
    #[error(transparent)]
    Install(#[source] miette::Report),
    #[error("could not determine default cache dir '{}'", .0.as_display())]
    CacheDir(miette::Report),
```

Reply:
I think the inner error will be enough.

Reply:
by doing so I have the following error:
```the method `as_dyn_error` exists for reference `&ErrReport`, but its trait bounds were not satisfied
the following trait bounds were not satisfied:```

I will keep it with additional message until we can disscus it in 1-1